legion/engine/llri-dx/source.cpp

```diff
@@ -9,88 +10,143 @@ namespace legion::graphics::llri
             }
 
-            if (extensionCreateResult != result::Success)
+            //Store user defined validation callback
+            //DirectX creates validation callbacks upon device creation so we just need to store information about this right now.
+            if (desc.callbackDesc.callback)
+            {
+                output->m_validationCallback = desc.callbackDesc;
+                output->m_validationCallbackMessenger = (void*)enableInternalAPIMessagePolling; //Use as boolean to indicate that a custom callback was set
+            }
+            else
+            {
+                output->m_validationCallback = { &internal::dummyValidationCallback, nullptr };
+                output->m_validationCallbackMessenger = (void*)0;
+            }
+
+            //Attempt to create factory
+            IDXGIFactory* factory = nullptr;
+            const HRESULT factoryCreateResult = CreateDXGIFactory2(factoryFlags, IID_PPV_ARGS(&factory));
+            if (FAILED(factoryCreateResult))
             {
-                destroyInstance(output);
-                return extensionCreateResult;
+                llri::destroyInstance(output);
+                return internal::mapHRESULT(factoryCreateResult);
             }
+
+            //Store factory and return result
+            output->m_ptr = factory;
+            *instance = output;
+            return result::Success;
         }
 
-        //Attempt to create factory
-        IDXGIFactory* factory = nullptr;
-        const HRESULT factoryCreateResult = CreateDXGIFactory2(factoryFlags, IID_PPV_ARGS(&factory));
-        if (FAILED(factoryCreateResult))
+        void impl_destroyInstance(Instance* instance)
         {
-            destroyInstance(output);
-            return internal::mapHRESULT(factoryCreateResult);
-        }
+            if (!instance)
+                return;
 
-        //Store factory and return result
-        output->m_ptr = factory;
-        *instance = output;
-        return result::Success;
-    }
+            for (auto& [ptr, adapter] : instance->m_cachedAdapters)
+                delete adapter;
 
-    void destroyInstance(Instance* instance)
-    {
-        if (!instance)
-            return;
+            if (instance->m_debugAPI)
+                static_cast<ID3D12Debug*>(instance->m_debugAPI)->Release();
 
-        for (auto& [ptr, adapter] : instance->m_cachedAdapters)
-            delete adapter;
+            if (instance->m_debugGPU)
+                static_cast<ID3D12Debug1*>(instance->m_debugGPU)->Release();
 
-        if (instance->m_debugAPI)
-            static_cast<ID3D12Debug*>(instance->m_debugAPI)->Release();
+            delete instance;
+        }
 
-        if (instance->m_debugGPU)
-            static_cast<ID3D12Debug1*>(instance->m_debugGPU)->Release();
+        void impl_pollAPIMessages(const validation_callback_desc& validation, void* messenger)
+        {
+            if (messenger != nullptr && messenger != (void*)1)
```

Comment:
Messenger is a bit of an awkward intermediate between the DirectX and Vulkan API:

DirectX allows the creation of an "InfoQueue", which enables you to poll messages from the API and use them. This is done during *Device* creation.

Vulkan allows the creation of a "DebugUtilsMessenger", which enables you to subscribe a callback for API messages. This is done during *Instance* creation.

Because these are both created at different times, their creation information needs to be available at different times. To deal with this, I specify the validation callback during llri::Instance creation, which is enough for Vulkan, but when llri::Device is created, I also pass that information on (hence why Instance::m_validationMessenger is set to a boolean during DirectX instance creation) to the Device so that DirectX can also create its structures appropriately.

When I poll messages, I assume the lowest possible level is relevant, so if llri::Device is available in a call I pass the device's messenger, and if it isn't then I fall back to the llri::Instance's messenger. In Vulkan's case this simply means the m_validationMessenger is passed down from Instance to Device, whereas in DirectX m_validationMessenger will be a boolean in Instance and an actual pointer in Device. 

Actually now that I'm giving it a second thought, I think it's better if I add an extra boolean that the messenger ought to be created, instead of trying to hide the information in m_validationMessenger. Would probably be a lot clearer than this system (and also resolves the pointer/boolref casting issue in one of the other conversations)

Reply:
hmmm okay i see what you're doing, tbh pretty neat. but yeah i'd either use a separate boolean or use `std::uintptr_t`